fix: Disconnect peers on handshake failure

diegomrsantos opened the pull request:
## Issue Addressed

https://github.com/sigp/anchor/issues/438

## Proposed Changes

- **Add explicit peer disconnection**: When a handshake fails (e.g., due to network mismatch), the network layer now explicitly disconnects the peer via `disconnect_peer_id()`
- **Improve disconnection timing**: Instead of relying on keep-alive timeouts (~10 seconds), peers are disconnected immediately upon handshake failure
- **Preserve existing test**: The handshake test continues to validate failure detection; disconnection behavior will be naturally tested through integration tests that use the full network stack
- **Add testing best practices**: Updated CLAUDE.md with guidance about not simulating production behavior in tests

## Technical Details

The change is made in `Network::handle_handshake_result()` where handshake failures are processed. When a handshake fails, we now call `self.swarm.disconnect_peer_id(peer_id)` to immediately disconnect the peer, rather than waiting for keep-alive timeouts.

This ensures faster and more deterministic peer disconnection on handshake failures, improving network behavior and reducing connection overhead.

dknopik commented:
I think this works because at some point there is a keepalive timeout as nothing is being sent anymore:

```
2025-07-21T09:33:38.548712Z DEBUG Swarm::poll: libp2p_swarm: Connection closed with error KeepAliveTimeout: Connected { endpoint: Listener { local_addr: /memory/4092149446297013635, send_back_addr: /memory/12899751130364311786 }, peer_id: PeerId("12D3KooWB1vGoK219R7Dnw7YUi4EcViLPWqjanY8F9P4TF27xL7u") } total_peers=0
2025-07-21T09:33:38.548789Z TRACE libp2p_swarm_test: event=ConnectionClosed { peer_id: PeerId("12D3KooWB1vGoK219R7Dnw7YUi4EcViLPWqjanY8F9P4TF27xL7u"), connection_id: ConnectionId(2), endpoint: Listener { local_addr: /memory/4092149446297013635, send_back_addr: /memory/12899751130364311786 }, num_established: 0, cause: Some(KeepAliveTimeout) }
2025-07-21T09:33:38.548814Z TRACE async_io::block_on: async_io::driver: sleep until notification
```

The test takes 10 seconds.

So I think #438 is actually valid.

We should modify this test to ensure the disconnection reason is correct (if possible), or ensure that the disconnection is soon after handshake failure.

diegomrsantos committed:
Disconnect peers on handshake failure

- Add explicit peer disconnection in network layer when handshake fails
- Update CLAUDE.md with testing best practices about not simulating production behavior in tests
- The existing handshake test validates failure detection; disconnection behavior will be tested through integration tests that use the full network stack

dknopik reviewed (dismissed):
Nice, thanks!

dknopik commented:
Sorry, can you please rebase this from `unstable` to `release-v0.3.0`?

diegomrsantos commented:
It seems to be good to be merged into `release-v0.3.0`

dknopik commented:
Currently, other changes are in this PR: [8f9cc96](https://github.com/sigp/anchor/pull/441/commits/8f9cc9693e6d4901511dbe3c4dbe61131e512685) and [17b8f05](https://github.com/sigp/anchor/pull/441/commits/17b8f0554ea8e21853006d7b5bbee8babfc1302f)

diegomrsantos commented:
> Currently, other changes are in this PR: [8f9cc96](https://github.com/sigp/anchor/pull/441/commits/8f9cc9693e6d4901511dbe3c4dbe61131e512685) and [17b8f05](https://github.com/sigp/anchor/pull/441/commits/17b8f0554ea8e21853006d7b5bbee8babfc1302f)

done

dknopik reviewed (approved):
LGTM, thank you!
